src/db.c

```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
+}
+
+static void
+queue_enum_end(struct db_queue_enum *queue_enum)
+{
+  if (!queue_enum->stmt)
+    return;
+
+  sqlite3_finalize(queue_enum->stmt);
+  sqlite3_free(queue_enum->filter);
+  queue_enum->filter = NULL;
+  queue_enum->stmt = NULL;
+}
+
+static inline char *
+strdup_if(char *str, int cond)
+{
+  if (str == NULL)
+    return NULL;
+
+  if (cond)
+    return strdup(str);
+
+  return str;
+}
+
+static int
+queue_enum_fetch(struct db_queue_enum *queue_enum, struct db_queue_item *queue_item, int keep_item)
+{
+  int ret;
+
+  memset(queue_item, 0, sizeof(struct db_queue_item));
+
+  if (!queue_enum->stmt)
+    {
+      DPRINTF(E_LOG, L_DB, "Queue enum not started!\n");
+      return -1;
+    }
+
+  ret = db_blocking_step(queue_enum->stmt);
+  if (ret == SQLITE_DONE)
+    {
+      DPRINTF(E_DBG, L_DB, "End of queue enum results\n");
+      return 0;
+    }
+  else if (ret != SQLITE_ROW)
+    {
+      DPRINTF(E_LOG, L_DB, "Could not step: %s\n", sqlite3_errmsg(hdl));
+      return -1;
+    }
+
+  queue_item->item_id = (uint32_t)sqlite3_column_int(queue_enum->stmt, 0);
+  queue_item->file_id = (uint32_t)sqlite3_column_int(queue_enum->stmt, 1);
+  queue_item->pos = (uint32_t)sqlite3_column_int(queue_enum->stmt, 2);
+  queue_item->shuffle_pos = (uint32_t)sqlite3_column_int(queue_enum->stmt, 3);
+  queue_item->data_kind = sqlite3_column_int(queue_enum->stmt, 4);
+  queue_item->media_kind = sqlite3_column_int(queue_enum->stmt, 5);
+  queue_item->song_length = (uint32_t)sqlite3_column_int(queue_enum->stmt, 6);
+  queue_item->path = strdup_if((char *)sqlite3_column_text(queue_enum->stmt, 7), keep_item);
+  queue_item->virtual_path = strdup_if((char *)sqlite3_column_text(queue_enum->stmt, 8), keep_item);
+  queue_item->title = strdup_if((char *)sqlite3_column_text(queue_enum->stmt, 9), keep_item);
+  queue_item->artist = strdup_if((char *)sqlite3_column_text(queue_enum->stmt, 10), keep_item);
+  queue_item->album_artist = strdup_if((char *)sqlite3_column_text(queue_enum->stmt, 11), keep_item);
+  queue_item->album = strdup_if((char *)sqlite3_column_text(queue_enum->stmt, 12), keep_item);
+  queue_item->genre = strdup_if((char *)sqlite3_column_text(queue_enum->stmt, 13), keep_item);
+  queue_item->songalbumid = sqlite3_column_int64(queue_enum->stmt, 14);
+  queue_item->time_modified = sqlite3_column_int(queue_enum->stmt, 15);
+  queue_item->artist_sort = strdup_if((char *)sqlite3_column_text(queue_enum->stmt, 16), keep_item);
+  queue_item->album_sort = strdup_if((char *)sqlite3_column_text(queue_enum->stmt, 17), keep_item);
+  queue_item->album_artist_sort = strdup_if((char *)sqlite3_column_text(queue_enum->stmt, 18), keep_item);
+  queue_item->year = sqlite3_column_int(queue_enum->stmt, 19);
+  queue_item->track = sqlite3_column_int(queue_enum->stmt, 20);
+  queue_item->disc = sqlite3_column_int(queue_enum->stmt, 21);
+
+  return 0;
+}
+
+int
+db_queue_enum_start(struct db_queue_enum *queue_enum)
+{
+  int ret;
+
+  db_transaction_begin();
```

Comment:
ok I guess the transactions are for locking, but why now - there was no real locking before in db.c, and I think it had similar read + write sequences